// backport/dist/lib/github/v4/fetchPullRequestAutoMergeMethod.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchPullRequestAutoMergeMethod = void 0;
const graphql_tag_1 = __importDefault(require("graphql-tag"));
const apiRequestV4_1 = require("./apiRequestV4");
async function fetchPullRequestAutoMergeMethod(options, pullNumber) {
    const { accessToken, githubApiBaseUrlV4, repoName, repoOwner } = options;
    const query = (0, graphql_tag_1.default) `
    query PullRequestAutoMergeMethod(
      $repoOwner: String!
      $repoName: String!
      $pullNumber: Int!
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        pullRequest(number: $pullNumber) {
          autoMergeRequest {
            enabledAt
            mergeMethod
          }
        }
      }
    }
  `;
    const res = await (0, apiRequestV4_1.apiRequestV4)({
        githubApiBaseUrlV4,
        accessToken,
        query,
        variables: {
            repoOwner,
            repoName,
            pullNumber,
        },
    });
    return res.repository.pullRequest?.autoMergeRequest?.mergeMethod;
}
exports.fetchPullRequestAutoMergeMethod = fetchPullRequestAutoMergeMethod;
