// backport/dist/lib/github/v4/FetchPullRequestId.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchPullRequestId = void 0;
const graphql_tag_1 = __importDefault(require("graphql-tag"));
const apiRequestV4_1 = require("./apiRequestV4");
async function fetchPullRequestId(options, pullNumber) {
    const { accessToken, githubApiBaseUrlV4, repoName, repoOwner } = options;
    const prQuery = (0, graphql_tag_1.default) `
    query PullRequestId(
      $repoOwner: String!
      $repoName: String!
      $pullNumber: Int!
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        pullRequest(number: $pullNumber) {
          id
        }
      }
    }
  `;
    const prResponse = await (0, apiRequestV4_1.apiRequestV4)({
        githubApiBaseUrlV4,
        accessToken,
        query: prQuery,
        variables: {
            repoOwner,
            repoName,
            pullNumber,
        },
    });
    return prResponse.repository.pullRequest.id;
}
exports.fetchPullRequestId = fetchPullRequestId;
