// backport/dist/lib/github/v4/fetchExistingPullRequest.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchExistingPullRequest = void 0;
const graphql_tag_1 = __importDefault(require("graphql-tag"));
const apiRequestV4_1 = require("./apiRequestV4");
async function fetchExistingPullRequest({ options, prPayload, }) {
    const { githubApiBaseUrlV4, accessToken } = options;
    const query = (0, graphql_tag_1.default) `
    query ExistingPullRequest(
      $repoOwner: String!
      $repoName: String!
      $base: String!
      $head: String!
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        name
        ref(qualifiedName: $head) {
          name
          associatedPullRequests(
            first: 1
            states: OPEN
            baseRefName: $base
            headRefName: $head
          ) {
            edges {
              node {
                number
                url
              }
            }
          }
        }
      }
    }
  `;
    const { repoForkOwner, head } = splitHead(prPayload);
    const res = await (0, apiRequestV4_1.apiRequestV4)({
        githubApiBaseUrlV4,
        accessToken,
        query,
        variables: {
            repoOwner: repoForkOwner,
            repoName: prPayload.repo,
            base: prPayload.base,
            head: head,
        },
    });
    const existingPullRequest = res.repository.ref?.associatedPullRequests.edges[0];
    if (!existingPullRequest) {
        return;
    }
    return {
        url: existingPullRequest.node.url,
        number: existingPullRequest.node.number,
    };
}
exports.fetchExistingPullRequest = fetchExistingPullRequest;
function splitHead(prPayload) {
    const [repoForkOwner, head] = prPayload.head.split(':');
    return { repoForkOwner, head };
}
