// cdk8s-cli/lib/import/k8s-util.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.safeParseJsonSchema = exports.parseApiTypeName = void 0;
const ajv_1 = __importDefault(require("ajv"));
const reviver_1 = require("../reviver");
const util_1 = require("../util");
var ApiLevel;
(function (ApiLevel) {
    ApiLevel["ALPHA"] = "alpha";
    ApiLevel["BETA"] = "beta";
    ApiLevel["STABLE"] = "stable";
})(ApiLevel || (ApiLevel = {}));
/**
 * Parses a fully qualified type name such as to it's components.
 */
function parseApiTypeName(fullname) {
    var _a, _b;
    const parts = fullname.split('.');
    const type = parts[parts.length - 1];
    const namespace = parts.slice(0, parts.length - 2).join('.');
    const prebase = parts[parts.length - 2];
    const version = /^v([0-9]+)(([a-z]+)([0-9]+))?$/.exec(prebase);
    return {
        fullname: fullname,
        version: version ? {
            raw: version[0],
            major: parseInt(version[1]),
            level: (_a = version[3]) !== null && _a !== void 0 ? _a : ApiLevel.STABLE,
            subversion: parseInt((_b = version[4]) !== null && _b !== void 0 ? _b : '0'),
        } : undefined,
        namespace: version ? namespace : `${namespace}.${prebase}`,
        basename: type,
    };
}
exports.parseApiTypeName = parseApiTypeName;
function safeParseJsonSchema(text) {
    const reviver = new reviver_1.SafeReviver({
        allowlistedKeys: ['$ref', '$schema'],
        sanitizers: [reviver_1.SafeReviver.DESCRIPTION_SANITIZER, reviver_1.SafeReviver.LEGAL_CHAR_SANITIZER],
    });
    const schema = (0, util_1.safeParseJson)(text, reviver);
    const ajv = new ajv_1.default();
    ajv.compile(schema);
    return schema;
}
exports.safeParseJsonSchema = safeParseJsonSchema;
