// backport/dist/lib/github/v3/addReviewersToPullRequest.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addReviewersToPullRequest = void 0;
const rest_1 = require("@octokit/rest");
const ora_1 = require("../../../lib/ora");
const logger_1 = require("../../logger");
async function addReviewersToPullRequest({ githubApiBaseUrlV3, repoName, repoOwner, accessToken, interactive, }, pullNumber, reviewers) {
    const text = `Adding reviewers: ${reviewers}`;
    logger_1.logger.info(text);
    const spinner = (0, ora_1.ora)(interactive, text).start();
    try {
        const octokit = new rest_1.Octokit({
            auth: accessToken,
            baseUrl: githubApiBaseUrlV3,
            log: logger_1.logger,
        });
        await octokit.pulls.requestReviewers({
            owner: repoOwner,
            repo: repoName,
            pull_number: pullNumber,
            reviewers,
        });
        spinner.succeed();
    }
    catch (e) {
        //@ts-expect-error
        const message = e.response?.data?.message;
        spinner.fail(`Adding reviewers. ${message ? message : ''}`);
        logger_1.logger.error(`Could not add reviewers to PR ${pullNumber}`, e);
    }
}
exports.addReviewersToPullRequest = addReviewersToPullRequest;
