// backport/dist/lib/github/v4/fetchCommits/fetchCommitByPullNumber.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fetchCommitByPullNumber = void 0;
const graphql_tag_1 = __importDefault(require("graphql-tag"));
const BackportError_1 = require("../../../BackportError");
const remoteConfig_1 = require("../../../remoteConfig");
const parseSourceCommit_1 = require("../../../sourceCommit/parseSourceCommit");
const apiRequestV4_1 = require("../apiRequestV4");
async function fetchCommitByPullNumber(options) {
    const { accessToken, githubApiBaseUrlV4 = 'https://api.github.com/graphql', pullNumber, repoName, repoOwner, } = options;
    const query = (0, graphql_tag_1.default) `
    query CommitByPullNumber(
      $repoOwner: String!
      $repoName: String!
      $pullNumber: Int!
    ) {
      repository(owner: $repoOwner, name: $repoName) {
        pullRequest(number: $pullNumber) {
          mergeCommit {
            ...SourceCommitWithTargetPullRequestFragment
          }
        }
      }
    }

    ${parseSourceCommit_1.SourceCommitWithTargetPullRequestFragment}
  `;
    let res;
    try {
        res = await (0, apiRequestV4_1.apiRequestV4)({
            githubApiBaseUrlV4,
            accessToken,
            query,
            variables: {
                repoOwner,
                repoName,
                pullNumber,
            },
        });
    }
    catch (e) {
        res = (0, remoteConfig_1.swallowMissingConfigFileException)(e);
    }
    const pullRequestNode = res.repository.pullRequest;
    if (!pullRequestNode) {
        throw new BackportError_1.BackportError(`The PR #${pullNumber} does not exist`);
    }
    const sourceCommit = pullRequestNode.mergeCommit;
    if (sourceCommit === null) {
        throw new BackportError_1.BackportError(`The PR #${pullNumber} is not merged`);
    }
    return (0, parseSourceCommit_1.parseSourceCommit)({ options, sourceCommit });
}
exports.fetchCommitByPullNumber = fetchCommitByPullNumber;
